### PR TITLE
ZEPPELIN-440 HiveInterpreter with multiple configuration

### PR DESCRIPTION
**Pull request by jongyoul**
This PR enable HiveInterpreter to configure several connection information. For using specific configurations, you should use some tricks like ``%hive(configuration_prefix)``. without it, you can use default settings starting with ``default``. 

screen shot 2015-11-21 at 1 51 31 am
screen shot 2015-11-21 at 1 53 59 am

### CODE REVIEW

**Comment by Leemoonsoo** [comment]
Like the idea about '('.

**Comment by jongyoul** [comment]
@Leemoonsoo Thanks. If you have any new or more Zeppelinous idea for specifying properties, feel free to talk. I'll make it pass to CI and add docs for this function.

**Comment by felixcheung** [comment]
It's nice. Would it be possible to make the interpreter group/name fully customizable?
For a generic JDBC interpreter, it would be great if one could set to `%hive` and another to `%presto`, or `%mysql`

**Comment by Leemoonsoo** [comment]
User can multiple settings (profiles) for an interpreter, in `Interpreter` menu. For example, 

spark-dev
spark-testing

settings can be made for the same %spark interpreter.

Current restriction is, only one interpreter setting can be used in the single notebook, in the same time.
That's because of selecting interpreter in paragraph is %[interpreter name], not %[setting name].

If we use %[setting name] for selecting interpreter,

Pros:
   * Allow multiple settings derived from single interpreter in a notebook.
   * User can set their favorite name for %[setting name] and use it in the paragraph

Cons:
   * When notebook file is shared to other people, it's hard to run while %[setting name] is customized and all Zeppelin instance will have all different setting name.

I think, in the future, we need to invent some way to allow all 'settings' being used in the same notebook at the same time, without having Cons that i mentioned above.

Until that time, I think this approach could be useful.

**Comment by felixcheung** [comment]
Good point - maybe we should save settings with notebook to make it portable?

**Comment by jongyoul** [comment]
@felixcheung First time, I also try to run interpreter by %{setting name}, but ... If it's possible to make notebook portable with settings, %{setting name} would be better.

**Comment by jongyoul** [comment]
trigger the CI again

**Comment by jongyoul** [comment]
travis fails while downloading dependencies

**Comment by jongyoul** [comment]
screen shot 2015-11-23 at 7 23 01 pm

**Comment by jongyoul** [comment]
@Leemoonsoo Could you please check the travis log? I think this is about ZeppelinSparkClusterTest. Do you have any idea?

**Comment by jongyoul** [comment]
this failure depends on #468

**Comment by jongyoul** [comment]
I revised HiveInterpreter.executeSql from PostgresSqlInterpreter.executeSql. I think that would be more general way to implement

**Comment by jongyoul** [comment]
@Leemoonsoo @felixcheung Please review this PR, again. I've almost done. Additionally, @Leemoonsoo, after review, can I try to merge this PR into apache master branch by myself? If it's possible, could you please tell me a procedure for it?

**Comment by bzz** [comment]
@jongyoul merge a PR is very simple, you just:
 - setup git remote, named `apache` pointing to original apache repo `https://git-wip-us.apache.org/repos/asf/incubator-zeppelin.git`
 - run [./dev/merge_zeppelin_pr.py](https://github.com/apache/incubator-zeppelin/blob/master/dev/merge_zeppelin_pr.py)
 - (bonus): if you have envs `JIRA_USERNAME ` and `JIRA_PASSWORD` setup - this script will take care of resolving the JIRA issue automatically

**Comment by Leemoonsoo** [comment]
About merging, you can use dev/merge_zeppelin_pr.py script.

I suggest clone zeppelin git repo into separate dir (separate from development) and then you'll need to add remote repo 'apache' and 'apache-github' to use this script.

```
apache	https://git-wip-us.apache.org/repos/asf/incubator-zeppelin (push)
apache	https://git-wip-us.apache.org/repos/asf/incubator-zeppelin (fetch)
apache-github	https://github.com/apache/incubator-zeppelin.git (fetch)
apache-github	https://github.com/apache/incubator-zeppelin.git (push)
```

Once it's done, just run dev/merge_zeppelin_pr.py will guide you merge the pull request.

File: hive/src/main/java/org/apache/zeppelin/hive/HiveInterpreter.java [hunk 1]
```diff
@@ -37,152 +40,294 @@
...
+    }
+    if (null == statement) {
+      statement = getConnection(propertyKey).createStatement();
+      paragraphIdStatementMap.put(paragraphId, statement);
+    }
+    return statement;
+  }
+
+  public InterpreterResult executeSql(String propertyKey,
+                                      String sql,
```
Comment (bzz) [inline hive/src/main/java/org/apache/zeppelin/hive/HiveInterpreter.java]: What do you think, does horizontal alignment makes it more readable or just makes it longer?

According to the styleguide Zeppelin project uses - it is never required and has some cons https://google.github.io/styleguide/javaguide.html#s4.6.3-horizontal-alignment

File: hive/src/main/java/org/apache/zeppelin/hive/HiveInterpreter.java [hunk 2]
```diff
@@ -6,9 +6,9 @@
  * to you under the Apache License, Version 2.0 (the
  * "License"); you may not use this file except in compliance
  * with the License.  You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
- *
+ * <p/>
```
Comment (bzz) [inline hive/src/main/java/org/apache/zeppelin/hive/HiveInterpreter.java]: This header now looks different from other one in this PR. 
Not sure if that is intentional and which one is right, but we better have only one

**Comment by bzz** [comment]
:+1: for test and docs!

File: hive/src/main/java/org/apache/zeppelin/hive/HiveInterpreter.java (hunk 2 above)
Comment (jongyoul) [inline hive/src/main/java/org/apache/zeppelin/hive/HiveInterpreter.java]: I'll revise this. My IDE setting rearranged this.

File: hive/src/main/java/org/apache/zeppelin/hive/HiveInterpreter.java (hunk 1 above)
Comment (jongyoul) [inline hive/src/main/java/org/apache/zeppelin/hive/HiveInterpreter.java]: Basically, I think it doesn't need except when that one line violates the maximum number of columns. I'll fix this, too.

**Comment by jongyoul** [comment]
@bzz Could you please verify this finally?

File: docs/interpreter/hive.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+%hive(etl)
+-- 'etl' is a ${prefix}
+select * from my_table;
+```
+
+You can also run multiple queries up to 10 by default. Changing these settings is not implemented yet.
+
+#### Apply Zeppelin Dynamic Forms
+
+You can leverage [Zepplein Dynamic Form](https://zeppelin.incubator.apache.org/docs/manual/dynamicform.html) inside your queries. You can use both the `text input` and `select form` parametrization features
```
Comment (felixcheung) [inline docs/interpreter/hive.md]: typo: Zeppelin

Comment (felixcheung) [inline docs/interpreter/hive.md]: parameterization

File: docs/interpreter/hive.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    <td><b>(Optional)</b>Password of the connection of `%hive(${prefix})`</td>
+  </tr>
+  <tr>
+    <td>${prefix}.xxx</td>
+    <td></td>
+    <td><b>(Optional)</b>Other properties used by the driver of `%hive(${prefix})`</td>
+  </tr>
+</table>
+
+This interpreter provides multiple configuration wih ${prefix}. User can set a multiple connection properties by this prefix. It can be used like `%hive(${prefix})`.
```
Comment (felixcheung) [inline docs/interpreter/hive.md]: typo: wih -> with

File: docs/interpreter/hive.md (hunk 3 above)
Comment (felixcheung) [inline docs/interpreter/hive.md]: also use this relative path for the link (url has changed for multiple releases)
`../manual/dynamicform.html`

Comment (Leemoonsoo) [inline docs/interpreter/hive.md]: Or {{BASE_PATH}}/manual/dynamicform.html would work

File: docs/interpreter/hive.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+%hive(etl)
+-- 'etl' is a ${prefix}
+select * from my_table;
+```
+
+You can also run multiple queries up to 10 by default. Changing these settings is not implemented yet.
+
+#### Apply Zeppelin Dynamic Forms
+
+You can leverage [Zeppelin Dynamic Form]({{BASE_PATH}}/manual/dynamicform.html) inside your queries. You can use both the `text input` and `select form` parameterization features
```
Comment (jongyoul) [inline docs/interpreter/hive.md]: @felixcheung Thanks for fixing them.
@Leemoonsoo Is is ok? I'm not good at writing docs.

Comment (Leemoonsoo) [inline docs/interpreter/hive.md]: Looks good

**Comment by Leemoonsoo** [comment]
You maybe also want to change a link https://github.com/apache/incubator-zeppelin/blob/master/docs/_includes/themes/zeppelin/_navigation.html#L41

**Comment by jongyoul** [comment]
@Leemoonsoo I've rebased and fixed it

**Comment by jongyoul** [comment]
Merging if there is no more discussion.

**Pull request closed without merging**